contracts/fuses/universal_token_swapper/UniversalTokenSwapperWithSignatureFuse.sol

```diff
@@ -0,0 +1,248 @@
+        _checkSubstrates(data_);
+
+        address plasmaVault = address(this);
+
+        Balances memory balances = Balances({
+            tokenInBalanceBefore: ERC20(data_.tokenIn).balanceOf(plasmaVault),
+            tokenOutBalanceBefore: ERC20(data_.tokenOut).balanceOf(plasmaVault),
+            tokenInBalanceAfter: 0,
+            tokenOutBalanceAfter: 0
+        });
+
+        if (data_.amountIn == 0) {
+            return;
+        }
+
+        ERC20(data_.tokenIn).safeTransfer(EXECUTOR, data_.amountIn);
+
+        SwapExecutorEth(EXECUTOR).execute(
+            SwapExecutorEthData({
+                tokenIn: data_.tokenIn,
+                tokenOut: data_.tokenOut,
+                targets: data_.data.targets,
+                callDatas: data_.data.callDatas,
+                ethAmounts: data_.data.ethAmounts,
+                tokensDustToCheck: data_.data.tokensDustToCheck
+            })
+        );
+
+        balances.tokenInBalanceAfter = ERC20(data_.tokenIn).balanceOf(plasmaVault);
+        balances.tokenOutBalanceAfter = ERC20(data_.tokenOut).balanceOf(plasmaVault);
+
+        if (balances.tokenInBalanceAfter >= balances.tokenInBalanceBefore) {
+            return;
+        }
+
+        uint256 tokenInDelta = balances.tokenInBalanceBefore - balances.tokenInBalanceAfter;
+
+        if (balances.tokenOutBalanceAfter <= balances.tokenOutBalanceBefore) {
+            revert UniversalTokenSwapperFuseSlippageFail();
+        }
+
+        uint256 tokenOutDelta = balances.tokenOutBalanceAfter - balances.tokenOutBalanceBefore;
+
+        address priceOracleMiddleware = PlasmaVaultLib.getPriceOracleMiddleware();
+
+        (uint256 tokenInPrice, uint256 tokenInPriceDecimals) = IPriceOracleMiddleware(priceOracleMiddleware)
+            .getAssetPrice(data_.tokenIn);
+        (uint256 tokenOutPrice, uint256 tokenOutPriceDecimals) = IPriceOracleMiddleware(priceOracleMiddleware)
+            .getAssetPrice(data_.tokenOut);
+
+        uint256 amountUsdInDelta = IporMath.convertToWad(
+            tokenInDelta * tokenInPrice,
+            IERC20Metadata(data_.tokenIn).decimals() + tokenInPriceDecimals
+        );
+        uint256 amountUsdOutDelta = IporMath.convertToWad(
+            tokenOutDelta * tokenOutPrice,
+            IERC20Metadata(data_.tokenOut).decimals() + tokenOutPriceDecimals
+        );
+
+        uint256 quotient = IporMath.division(amountUsdOutDelta * 1e18, amountUsdInDelta);
+
+        if (quotient < SLIPPAGE_REVERSE) {
+            revert UniversalTokenSwapperFuseSlippageFail();
+        }
+
+        _emitUniversalTokenSwapperFuseEnter(data_, tokenInDelta, tokenOutDelta);
+    }
+
+    /// @notice Converts UniversalTokenSwapperSubstrate to bytes32
+    /// @param substrate_ The substrate to convert
+    /// @return The packed bytes32 representation
+    function toBytes32(UniversalTokenSwapperSubstrate memory substrate_) public pure returns (bytes32) {
```

Comment:
why public?

Reply:
for the frontend to be able to use this for data encoding